src/Analysis/Ast/Impl/Specializations/BuiltinsSpecializations.cs

```diff
@@ -91,6 +93,79 @@ public static IMember List(IPythonInterpreter interpreter, IPythonFunctionOverlo
+                if (classType?.Mro == null) {
+                    return null;
+                }
+
+                // second argument optional
+                bool isUnbound = argSet.Argument<IMember>(1) == null;
+                if (isUnbound) {
+                    return CreateSuper(argSet, classType.Mro);
+                } else {
+                    // Second argument can be either an object instance or a class which are both IPythonType
+                    IPythonType objOrType = null;
+                    objOrType = argSet.Argument<IMember>(1).GetPythonType();
+                    bool isInstance = objOrType?.Equals(classType) ?? false;
+
+                    if (!isInstance) {
+                        var argAsType = argSet.Argument<IMember>(1).GetPythonType<IPythonClassType>();
+                        bool isSubClass = classType?.Bases.Any(b => argAsType.Equals(b)) ?? false;
+                        if (isSubClass) {
+                            objOrType = argAsType as IPythonType;
+                        }
+                    }
+                    
+                    return (objOrType != null) ? CreateSuper(argSet, classType.Mro, newStartType: objOrType) : null;
+                }
+            } else {
+                //Zero argument form only works inside a class definition
+                foreach (var s in argSet.Eval.CurrentScope.EnumerateTowardsGlobal) {
+                    if (s.Node is ClassDefinition) {
+                        var classType = s.Variables["__class__"].GetPythonType<IPythonClassType>();
+
+                        if (classType?.Mro != null) {
+                            return CreateSuper(argSet, classType.Mro);
+                        }
+                    }
+                }
+            }
+            return null;
+        }
+
+        private static IMember CreateSuper(IArgumentSet argSet, IReadOnlyList<IPythonType> callerMro, IPythonType newStartType = null) {
+            if (!callerMro.Any()) {
+                return null;
+            }
+
+            // common case super with no params just reuse the caller's readonly mro
+            IReadOnlyList<IPythonType> mro = callerMro;
+            
+            // skip doing work if the newStartType is the first element in the callers mro
+            if (newStartType != null &&
+                !newStartType.Equals(callerMro.FirstOrDefault())) {
+                var mroList = callerMro.ToList();
+                var start = mroList.FindIndex(0, t =>  t.Equals(newStartType));
+                if (start >= 0) {
+                    mro = mroList.GetRange(start, mro.Count() - start).ToArray();
+                } else {
+                    return null;  // newStartType wasn't valid
+                }
+            }
+            
+            if (mro.Any()) {
+                var nextClassInLine = mro.FirstOrDefault();
+                var location = new Location(nextClassInLine.Location.Module, nextClassInLine.Location.IndexSpan);
```

Comment:
Double enumeration. `mro` is list so just check `Count` or do `var nextClassInLine = mro.FirstOrDefault()` and then check `nextClassInLine` for null.

Reply:
ya i had the check ```nextClassInLine``` for null first then removed it

Reply:
i'll add it back